### PR TITLE
Add basic 3D symmetry (X axis)

### PR DESCRIPTION
**Pull request by CI09**
**Brief Description of What This PR Does**

This PR does some stuff...

**Related Issues**

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Review by dligr — COMMENTED** [review]
I tested the changes to work in game. The only issue seems to be that metaballs can't be removed using the context menu. I think that this is due to the fact that context deletion is done by `OnDeletePressed()`, which doesn't use `RunWithSymmetry()`.

File: src/late_multicellular_stage/editor/MetaballEditorComponentBase.cs [hunk 1]
```diff
@@ -708,8 +709,8 @@ protected void GetMouseMetaball(out Vector3 position, out TMetaball? metaball, f
     ///     TODO: this is not implemented currently and just returns the given primary position
```
Comment (dligr) [inline src/late_multicellular_stage/editor/MetaballEditorComponentBase.cs:709]: Since the function is now implemented, I think that this comment should be updated

File: src/late_multicellular_stage/editor/MetaballEditorComponentBase.cs [hunk 2]
```diff
@@ -721,6 +722,34 @@ protected void RunWithSymmetry(Vector3 position, TMetaball? parent, Action<Vecto
...
+                    {
+                        callback(symmetryPosition, parent);
+                    }
+                }
+
+                break;
+            }
+
             default:
                 throw new NotSupportedException("symmetry editing not implemented yet");
```
Comment (dligr) [inline src/late_multicellular_stage/editor/MetaballEditorComponentBase.cs:754]: Also, I think that this error message should be updated to mean that some symmetry types aren't supported, as opposed to symmetry not being implemented

File: src/late_multicellular_stage/MetaballLayout.cs [hunk 3]
```diff
@@ -209,4 +210,26 @@ public IEnumerable<T> GetMetaballsNotTouchingParents(float contactThreshold = 0.
             }
         }
     }
+
+    public Metaball? GetClosestMetaballToPosition(Vector3 position)
+    {
+        var closestDistance = 10000.0f;
```
Comment (hhyyrylainen) [inline src/late_multicellular_stage/MetaballLayout.cs:216]: This should default to `float.Max` so that this always works.

File: src/late_multicellular_stage/MetaballLayout.cs [hunk 4]
```diff
@@ -209,4 +210,26 @@ public IEnumerable<T> GetMetaballsNotTouchingParents(float contactThreshold = 0.
...
+        Metaball? closestMetaball = null;
+
+        foreach (var metaball in this)
+        {
+            if (metaball.Parent == null)
+                continue;
+
+            var distance = metaball.Position.DistanceTo(position);
```
Comment (hhyyrylainen) [inline src/late_multicellular_stage/MetaballLayout.cs:224]: To save massively on performance this should use squared distances. (as calculating the distance requires taking a square root).

File: src/late_multicellular_stage/editor/MetaballEditorComponentBase.cs [hunk 5]
```diff
@@ -721,8 +717,36 @@ protected void RunWithSymmetry(Vector3 position, TMetaball? parent, Action<Vecto
...
+                {
+                    callback(position, parent);
+                }
+                else
+                {
+                    callback(position, parent);
+
+                    if (parent != null)
+                    {
+                        var symmetryParent = (TMetaball?)editedMetaballs.GetClosestMetaballToPosition(
```
Comment (hhyyrylainen) [inline src/late_multicellular_stage/editor/MetaballEditorComponentBase.cs:734]: This seems like a place that could be done more type safely.

I think I'll have time tomorrow to fix up the few small things left in this if they aren't done by then yet.

File: src/late_multicellular_stage/MulticellularMetaball.cs [hunk 6]
```diff
@@ -14,7 +14,7 @@ public MulticellularMetaball(CellType cellType)
     ///   The cell type this metaball consists of
     /// </summary>
     [JsonProperty]
-    public CellType CellType { get; private set; }
+    public CellType CellType { get; set; }
```
Comment (hhyyrylainen) [inline src/late_multicellular_stage/MulticellularMetaball.cs:17]: This seems like an unnecessary change for this PR?

**Comment by CI09** [comment]
This symmetry is kinda primitive so I would appreciate if in the release it is mentioned that this is basic and better will be yet to be implemented

**Comment by hhyyrylainen** [comment]
This is not going to be one of the headline features. If I remember I'll call this "Added basic macroscopic editor symmetry mode" in the release notes.

File: src/late_multicellular_stage/MetaballLayout.cs [hunk 7]
```diff
@@ -209,4 +210,26 @@ public IEnumerable<T> GetMetaballsNotTouchingParents(float contactThreshold = 0.
...
     }
+
+    public Metaball? GetClosestMetaballToPosition(Vector3 position)
+    {
+        var closestDistance = float.MaxValue;
+        Metaball? closestMetaball = null;
+
+        foreach (var metaball in this)
+        {
+            if (metaball.Parent == null)
```
Comment (hhyyrylainen) [inline src/late_multicellular_stage/MetaballLayout.cs:221]: I have one question, why cannot this return the root metaball? Sometimes that is closest to the specified position, right?

Comment (CI09) [reply src/late_multicellular_stage/MetaballLayout.cs:221]: You're right

File: src/late_multicellular_stage/editor/MetaballEditorComponentBase.cs (hunk 5 above)
Comment (hhyyrylainen) [reply src/late_multicellular_stage/editor/MetaballEditorComponentBase.cs:734]: I'll commit a fix for this, was just a case of using `T` as the return type in the generic type to return the real metaball type to remove the need for this cast.

**Review by hhyyrylainen — APPROVED** [review]
I did a few more code tweaks and fixes, and also tested in game. I think this is now good to go to get into the RC I'll make soon.

**Comment by CI09** [comment]
FYI metaball resize branch is ready right now.

**Comment by hhyyrylainen** [comment]
I'll make the RC build in roughly an hour so there's still a chance for that to make it in as long as the changes are really simple and quick to review.
